File: VB/XPO/ExtendedEvents.Module/BusinessObjects/ExtendedEvent.vb
Imports DevExpress.Persistent.Base
Imports DevExpress.Persistent.BaseImpl
Imports DevExpress.Xpo

Namespace ExtendedEvents.Module.BusinessObjects

    <DefaultClassOptions>
    Public Class ExtendedEvent
        Inherits [Event]

        Public Sub New(ByVal session As Session)
            MyBase.New(session)
        End Sub

        Public Property CustomSimpleTypeField As String
            Get
                Return GetPropertyValue(Of String)(NameOf(ExtendedEvent.CustomSimpleTypeField))
            End Get

            Set(ByVal value As String)
                SetPropertyValue(NameOf(ExtendedEvent.CustomSimpleTypeField), value)
            End Set
        End Property

        Public Property CustomReferenceTypeField As CustomReferenceTypeField
            Get
                Return GetPropertyValue(Of CustomReferenceTypeField)(NameOf(ExtendedEvent.CustomReferenceTypeField))
            End Get

            Set(ByVal value As CustomReferenceTypeField)
                SetPropertyValue(NameOf(ExtendedEvent.CustomReferenceTypeField), value)
            End Set
        End Property
    End Class

    Public Class CustomReferenceTypeField
        Inherits BaseObject

        Public Sub New(ByVal session As Session)
            MyBase.New(session)
        End Sub

        Public Property Name As String
            Get
                Return GetPropertyValue(Of String)(NameOf(CustomReferenceTypeField.Name))
            End Get

            Set(ByVal value As String)
                SetPropertyValue(NameOf(CustomReferenceTypeField.Name), value)
            End Set
        End Property
    End Class
End Namespace

File: CS/XPO/ExtendedEvents.Module/BusinessObjects/ExtendedEvent.cs
using DevExpress.Persistent.Base;
using DevExpress.Persistent.BaseImpl;
using DevExpress.Persistent.Validation;
using DevExpress.Xpo;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ExtendedEvents.Module.BusinessObjects;
[DefaultClassOptions]
public class ExtendedEvent : Event {
    public ExtendedEvent(Session session) : base(session) {
    }
    public string CustomSimpleTypeField {
        get => GetPropertyValue<string>(nameof(CustomSimpleTypeField));
        set => SetPropertyValue(nameof(CustomSimpleTypeField), value);
    }
    public CustomReferenceTypeField CustomReferenceTypeField {
        get => GetPropertyValue<CustomReferenceTypeField>(nameof(CustomReferenceTypeField));
        set => SetPropertyValue(nameof(CustomReferenceTypeField), value);
    }
}
public class CustomReferenceTypeField : BaseObject {
    public CustomReferenceTypeField(Session session) : base(session) {
    }
    public string Name {
        get => GetPropertyValue<string>(nameof(Name));
        set => SetPropertyValue(nameof(Name), value);
    }
}
